docs/tutorials/01-instalacao-configuracao.md
# Tutorial 1: Instalação e Configuração Inicial do Polis

## 📖 Objetivos
- Instalar o Polis no seu sistema
- Configurar o ambiente de desenvolvimento
- Verificar a instalação
- Entender a estrutura básica do projeto

## ⏱️ Duração
- **Estimada**: 15 minutos
- **Nível**: Iniciante

## 📋 Pré-requisitos
- Sistema operacional Linux, macOS ou Windows
- Rust 1.70+ instalado
- Git instalado
- Acesso à internet

## 🎬 Roteiro do Vídeo

### Introdução (0:00 - 2:00)
- Apresentação do Polis
- O que é um container runtime
- Por que escolher o Polis
- O que será coberto neste tutorial

### Instalação do Rust (2:00 - 5:00)
```bash
# Instalar Rust via rustup
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
source ~/.cargo/env

# Verificar instalação
rustc --version
cargo --version
```

### Clonagem do Repositório (5:00 - 7:00)
```bash
# Clonar o repositório
git clone https://github.com/polis-project/polis.git
cd polis

# Verificar estrutura do projeto
ls -la
```

### Compilação do Projeto (7:00 - 10:00)
```bash
# Compilar o projeto
cargo build --release

# Verificar se compilou corretamente
ls target/release/
```

### Configuração Inicial (10:00 - 12:00)
```bash
# Criar diretório de configuração
mkdir -p ~/.polis

# Criar arquivo de configuração básico
cat > ~/.polis/config.yaml << EOF
storage:
  root_dir: ~/.polis/data
  image_cache_dir: ~/.polis/images

network:
  bridge_name: polis0
  subnet: "172.17.0.0/16"

api:
  rest_port: 8080
  grpc_port: 9090

logging:
  level: info
  file: ~/.polis/logs/polis.log
EOF
```

### Verificação da Instalação (12:00 - 15:00)
```bash
# Executar o CLI
./target/release/polis --version

# Verificar comandos disponíveis
./target/release/polis --help

# Testar criação de container
./target/release/polis container create --name test --image alpine:latest
```

## 💻 Código de Exemplo

### Script de Instalação Automática
```bash
#!/bin/bash
# install_polis.sh

set -e

echo "🚀 Instalando Polis..."

# Verificar Rust
if ! command -v rustc &> /dev/null; then
    echo "📦 Instalando Rust..."
    curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
    source ~/.cargo/env
fi

# Clonar repositório
if [ ! -d "polis" ]; then
    echo "📥 Clonando repositório..."
    git clone https://github.com/polis-project/polis.git
fi

cd polis

# Compilar
echo "🔨 Compilando Polis..."
cargo build --release

# Criar configuração
echo "⚙️ Criando configuração..."
mkdir -p ~/.polis
cp examples/config.yaml ~/.polis/

echo "✅ Instalação concluída!"
echo "Execute: ./target/release/polis --help"
```

### Configuração Avançada
```yaml
# ~/.polis/config.yaml
storage:
  root_dir: ~/.polis/data
  image_cache_dir: ~/.polis/images
  volume_dir: ~/.polis/volumes

network:
  bridge_name: polis0
  subnet: "172.17.0.0/16"
  gateway: "172.17.0.1"
  dns_servers:
    - "8.8.8.8"
    - "8.8.4.4"

api:
  rest_port: 8080
  grpc_port: 9090
  host: "0.0.0.0"

security:
  enable_apparmor: true
  enable_selinux: true
  default_capabilities:
    - "NET_ADMIN"
    - "SYS_ADMIN"

logging:
  level: info
  file: ~/.polis/logs/polis.log
  max_size: "100MB"
  max_files: 5

monitoring:
  enable_metrics: true
  metrics_port: 9091
  health_check_interval: "30s"
```

## 🔗 Recursos Adicionais
- [Documentação de Instalação](../INSTALACAO.md)
- [Configuração Avançada](../CONFIGURACAO.md)
- [Troubleshooting](../TROUBLESHOOTING.md)
- [Fórum da Comunidade](https://github.com/polis-project/polis/discussions)

## ❓ Exercícios
1. **Instalação Básica**: Siga o tutorial e instale o Polis
2. **Configuração Personalizada**: Modifique o arquivo de configuração
3. **Verificação**: Execute todos os comandos de verificação
4. **Exploração**: Use `polis --help` para ver todos os comandos

## 🎯 Próximos Tutoriais
- [Tutorial 2: Primeiros Passos com Containers](02-primeiros-passos.md)
- [Tutorial 3: Conceitos Fundamentais](03-conceitos-fundamentais.md)

## 📝 Notas do Instrutor

### Pontos Importantes
- Sempre verificar a versão do Rust antes de começar
- Explicar a diferença entre `cargo build` e `cargo build --release`
- Mostrar como interpretar mensagens de erro comuns
- Enfatizar a importância da configuração inicial

### Dicas de Apresentação
- Usar uma tela limpa e organizada
- Falar devagar e pausar entre comandos
- Explicar o que cada comando faz
- Mostrar a saída esperada de cada comando

### Possíveis Problemas
- **Erro de compilação**: Verificar versão do Rust
- **Permissões**: Usar `sudo` se necessário
- **Rede**: Verificar conectividade para clonar repositório
- **Espaço**: Verificar espaço em disco disponível

### Tempo Sugerido por Seção
- Introdução: 2 min
- Instalação Rust: 3 min
- Clonagem: 2 min
- Compilação: 3 min
- Configuração: 2 min
- Verificação: 3 min
- **Total**: 15 min


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.1.0] - 2024-01-15

### Added
- Initial release of Polis Container Runtime
- Complete container lifecycle management
- OCI image support
- REST and gRPC APIs
- Security features (namespaces, cgroups, seccomp, AppArmor, SELinux)
- Network management with load balancing
- Storage management with volumes
- Monitoring and observability
- Authentication and authorization
- CLI interface
- Service discovery and orchestration
- Auto-scaling capabilities
- Health monitoring system
- Performance optimization
- Memory management improvements
- Comprehensive documentation
- CI/CD pipeline
- Performance benchmarks
- Video tutorials and community guidelines

### Features
- Container creation, start, stop, pause, resume
- Image registry integration
- Advanced networking with port forwarding
- Persistent storage with bind mounts
- Service discovery and load balancing
- Auto-scaling based on metrics
- Health checks and monitoring
- REST and gRPC APIs
- Authentication with JWT
- Comprehensive CLI
- Docker-compatible images
- Multi-platform support (Linux, macOS, Windows)

### Technical Details
- Built in Rust for safety and performance
- Async/await throughout
- Modular architecture
- OCI-compliant
- Production-ready
- Enterprise features

### Installation
- Pre-built binaries for all platforms
- Package manager support
- Docker images
- Source code compilation

### Documentation
- User guide
- API reference
- Installation guide
- Developer guide
- Video tutorials

### Support
- GitHub Issues
- Discord Community
- Comprehensive documentation

### License
MIT License

polis-core/src/config.rs
use crate::error::{PolisError, Result};
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct PolisConfig {
    pub runtime: RuntimeConfig,
    pub storage: StorageConfig,
    pub network: NetworkConfig,
    pub security: SecurityConfig,
    pub api: ApiConfig,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RuntimeConfig {
    pub root_dir: PathBuf,
    pub log_level: LogLevel,
    pub debug: bool,
    pub max_containers: u32,
    pub container_timeout: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StorageConfig {
    pub driver: StorageDriver,
    pub root_dir: PathBuf,
    pub max_size: Option<u64>,
    pub cleanup_on_exit: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NetworkConfig {
    pub driver: NetworkDriver,
    pub bridge_name: String,
    pub subnet: Option<String>,
    pub gateway: Option<String>,
    pub dns_servers: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SecurityConfig {
    pub seccomp_profile: String,
    pub apparmor_profile: String,
    pub no_new_privileges: bool,
    pub drop_capabilities: Vec<String>,
    pub read_only_rootfs: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiConfig {
    pub rest_port: u16,
    pub grpc_port: u16,
    pub host: String,
    pub enable_cors: bool,
    pub timeout_seconds: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum LogLevel {
    Error,
    Warn,
    Info,
    Debug,
    Trace,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum StorageDriver {
    Overlay2,
    Btrfs,
    Zfs,
    Dir,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum NetworkDriver {
    Bridge,
    Host,
    None,
    Macvlan,
}

impl Default for RuntimeConfig {
    fn default() -> Self {
        Self {
            root_dir: PathBuf::from("/var/lib/polis"),
            log_level: LogLevel::Info,
            debug: false,
            max_containers: 100,
            container_timeout: 30,
        }
    }
}

impl Default for StorageConfig {
    fn default() -> Self {
        Self {
            driver: StorageDriver::Overlay2,
            root_dir: PathBuf::from("/var/lib/polis/storage"),
            max_size: None,
            cleanup_on_exit: true,
        }
    }
}

impl Default for NetworkConfig {
    fn default() -> Self {
        Self {
            driver: NetworkDriver::Bridge,
            bridge_name: "polis0".to_string(),
            subnet: Some("172.17.0.0/16".to_string()),
            gateway: Some("172.17.0.1".to_string()),
            dns_servers: vec!["8.8.8.8".to_string(), "8.8.4.4".to_string()],
        }
    }
}

impl Default for SecurityConfig {
    fn default() -> Self {
        Self {
            seccomp_profile: "default".to_string(),
            apparmor_profile: "docker-default".to_string(),
            no_new_privileges: true,
            drop_capabilities: vec!["ALL".to_string()],
            read_only_rootfs: false,
        }
    }
}

impl Default for ApiConfig {
    fn default() -> Self {
        Self {
            rest_port: 8080,
            grpc_port: 9090,
            host: "0.0.0.0".to_string(),
            enable_cors: true,
            timeout_seconds: 30,
        }
    }
}

impl PolisConfig {
    pub fn load_from_file(path: &str) -> Result<Self> {
        let content = std::fs::read_to_string(path).map_err(PolisError::Io)?;

        let config = if path.ends_with(".yaml") || path.ends_with(".yml") {
            serde_yaml::from_str(&content)
                .map_err(|e| PolisError::Runtime(format!("Erro ao parsear YAML: {}", e)))?
        } else if path.ends_with(".toml") {
            toml::from_str(&content)
                .map_err(|e| PolisError::Runtime(format!("Erro ao parsear TOML: {}", e)))?
        } else if path.ends_with(".json") {
            serde_json::from_str(&content).map_err(PolisError::Serialization)?
        } else {
            return Err(PolisError::Runtime(format!(
                "Formato de arquivo não suportado: {}",
                path
            )));
        };

        Ok(config)
    }

    pub fn save_to_file(&self, path: &str) -> Result<()> {
        let content = if path.ends_with(".yaml") || path.ends_with(".yml") {
            serde_yaml::to_string(self)
                .map_err(|e| PolisError::Runtime(format!("Erro ao serializar YAML: {}", e)))?
        } else if path.ends_with(".toml") {
            toml::to_string(self)
                .map_err(|e| PolisError::Runtime(format!("Erro ao serializar TOML: {}", e)))?
        } else if path.ends_with(".json") {
            serde_json::to_string_pretty(self).map_err(PolisError::Serialization)?
        } else {
            return Err(PolisError::Runtime(format!(
                "Formato de arquivo não suportado: {}",
                path
            )));
        };

        std::fs::write(path, content).map_err(PolisError::Io)?;

        Ok(())
    }

    pub fn validate(&self) -> Result<()> {
        if self.runtime.max_containers == 0 {
            return Err(PolisError::Runtime(
                "max_containers deve ser maior que 0".to_string(),
            ));
        }

        if self.runtime.container_timeout == 0 {
            return Err(PolisError::Runtime(
                "container_timeout deve ser maior que 0".to_string(),
            ));
        }

        if self.api.rest_port == 0 || self.api.grpc_port == 0 {
            return Err(PolisError::Runtime(
                "Portas da API devem ser maiores que 0".to_string(),
            ));
        }

        if self.api.rest_port == self.api.grpc_port {
            return Err(PolisError::Runtime(
                "Portas REST e gRPC devem ser diferentes".to_string(),
            ));
        }

        Ok(())
    }
}
